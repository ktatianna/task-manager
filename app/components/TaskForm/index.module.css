.form {
  align-items: center;
  background: #fff;
  border-bottom: 1px solid #e5e7eb;
  box-sizing: border-box;
  display: flex;
  margin-bottom: 16px;
  padding: 16px;
  width: 400px;
}

.input {
  border-color: #e5e7eb;
  border-radius: 8px;
  border-width: 2px;
  flex: 1;
  margin: 0 16px 0 0;
  outline: 2px solid transparent;
  outline-offset: 2px;
  padding: 8px;
}

.input:focus {
  border-color: #3b82f6;
}

.button {
  background-color: #a851f9da;
  border-radius: 8px;
  color: white;
  cursor: pointer;
  padding: 8px 20px;
  text-align: center;
}

.button:hover {
  background-color: #8a27e5da;
}