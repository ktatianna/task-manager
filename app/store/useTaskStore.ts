import { create } from "zustand";
import { Task } from "../types/task";
import { v4 as uuidv4 } from "uuid";

interface TaskStore {
  tasks: Task[];
  addTask: (description: string) => void;
  toggleTask: (id: string) => void;
  deleteTask: (id: string) => void;
  clearCompleted: () => void;
}

export const useTaskStore = create<TaskStore>((set) => ({
  tasks: [],
  addTask: (description) =>
    set((state) => ({
      tasks: [...state.tasks, { id: uuidv4(), description, completed: false }],
    })),
  toggleTask: (id) =>
    set((state) => ({
      tasks: state.tasks.map((task) =>
        task.id === id ? { ...task, completed: !task.completed } : task
      ),
    })),
  deleteTask: (id) =>
    set((state) => ({
      tasks: state.tasks.filter((task) => task.id !== id),
    })),
  clearCompleted: () =>
    set((state) => ({
      tasks: state.tasks.filter((task) => !task.completed),
    })),
}));
