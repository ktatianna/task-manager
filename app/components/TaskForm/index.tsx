"use client";

import { useState } from "react";
import { useTaskStore } from "@/app/store/useTaskStore";

import styles from "./index.module.css";

const TaskForm = () => {
  const [description, setDescription] = useState("");
  const addTask = useTaskStore((state) => state.addTask);

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (description.trim() === "") {
      return;
    }
    addTask(description.trim());
    setDescription("");
  };

  return (
    <form onSubmit={handleSubmit} className={styles.form}>
      <input
        className={styles.input}
        type="text"
        value={description}
        onChange={(e) => setDescription(e.target.value)}
        placeholder="Add a new task"
        aria-label="Add a new task"
      />
      <button
        className={styles.button}
        type="submit"
      >
        Add
      </button>
    </form>
  );
};

export default TaskForm;
