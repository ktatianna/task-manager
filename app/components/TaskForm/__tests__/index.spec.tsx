import { render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event';
import { useTaskStore } from "@/app/store/useTaskStore";
import TaskForm from "../index";

jest.mock("@/app/store/useTaskStore", () => ({
  useTaskStore: jest.fn(),
}));

const setup = (propsOverride = {}) => {
  const props = { ...propsOverride };

  return {
    ...render(<TaskForm {...props} />),
    props,
  };
};

describe("TaskForm", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should match the snapshot", () => {
    (useTaskStore as unknown as jest.Mock).mockReturnValue(() => {});
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should update the input value when typing", async () => {
    (useTaskStore as unknown as jest.Mock).mockReturnValue(() => {});
    setup();
    const input = screen.getByPlaceholderText("Add a new task") as HTMLInputElement;
    await userEvent.type(input, "Nueva tarea");
    expect(input.value).toBe("Nueva tarea");
  });

  it("should call addTask and clear input when submitting a non-empty value", async () => {
    const addTaskMock = jest.fn();
    (useTaskStore as unknown as jest.Mock).mockReturnValue(addTaskMock);

    setup();
    const input = screen.getByPlaceholderText("Add a new task") as HTMLInputElement;
    const AddButton = screen.getByRole("button", { name: "Add" });

    await userEvent.type(input, "Nueva tarea");
    await userEvent.click(AddButton);

    expect(addTaskMock).toHaveBeenCalledWith("Nueva tarea");
    expect(input.value).toBe("");
  });

  it("should not call addTask if input is empty", async () => {
    const addTaskMock = jest.fn();
    (useTaskStore as unknown as jest.Mock).mockReturnValue(addTaskMock);

    setup();
    const button = screen.getByRole("button", { name: "Add" });

    await userEvent.click(button);

    expect(addTaskMock).not.toHaveBeenCalled();
  });
});