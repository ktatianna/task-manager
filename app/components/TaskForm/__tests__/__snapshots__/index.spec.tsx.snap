// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm should match the snapshot 1`] = `
<div>
  <form
    class="form"
  >
    <input
      aria-label="Add a new task"
      class="input"
      placeholder="Add a new task"
      type="text"
      value=""
    />
    <button
      class="button"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
